jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java

```diff
@@ -1379,12 +1362,89 @@ private void shutdownSessions()
     {
         _sessionCache.shutdown();
     }
-    
-    public record RequestedSession(ManagedSession session, String sessionId, boolean sessionIdFromCookie)
-    {
-    }
 
-    private static final  RequestedSession NO_REQUESTED_SESSION = new RequestedSession(null, null, false);
+    /**
+     * Details of the requested session.
+     * Session implementations should make an instance of this record available as a hidden (not in name set) request
+     * attribute for the name "org.eclipse.jetty.session.AbstractSessionManager$RequestedSession"
+     * @param session The {@link Session} associated with the ID, which may have been invalidated or changed ID since the
+     *                request was received; or {@code null} if no session existed matching the requested ID.
+     * @param sessionId The requested session ID.
+     * @param sessionIdFrom A {@link String} representing the source of the session ID.  Common values include:
+     *                      {@link #ID_FROM_COOKIE} or {@link #ID_FROM_URI_PARAMETER} if there is no ID.
+     */
+    public record RequestedSession(ManagedSession session, String sessionId, String sessionIdFrom)
+    {
+        public static final RequestedSession NO_REQUESTED_SESSION = new RequestedSession(null, null, null);
+        public static final String ATTRIBUTE = "org.eclipse.jetty.session.RequestedSession";
+        public static final String ID_FROM_COOKIE = "cookie";
+        public static final String ID_FROM_URI_PARAMETER = "uri";
+
+        /**
+         * Get the {@code RequestedSession} by attribute
+         * @param request The attributes to query
+         * @return The found {@code RequestedSession} or {@link #NO_REQUESTED_SESSION} if none found. Never {@code null}.
+         */
+        public static RequestedSession byAttribute(Attributes request)
```

Comment:
If we have the ability to retrieve the information via individual attributes, why do we need to expose the `RequestedSession` record at all? I'm not really sure it should be exposed, it was only ever intended as an information carrier for the internal implementation. Also, shouldn't the attributes have constant definitions?

Reply:
It was already exposed to some extent to anything extending AbstractSessionManager.  It is used by both core and servlet session handlers.